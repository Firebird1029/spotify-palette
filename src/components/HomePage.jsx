/* eslint-disable no-console */
/* eslint-disable react/prop-types */

"use client";

import axios from "axios";
import React, { useRef, useState, useEffect } from "react";

function HomeContent() {
  const [token, setToken] = useState(null);
  const [userProfile, setUserProfile] = useState(null);
  const [topArtists, setTopArtists] = useState([]);

  const CLIENT_ID = "319f3f19b0794ac28b1df51ca946609c";
  const REDIRECT_URI = "http://localhost:3000";
  const AUTH_ENDPOINT = "https://accounts.spotify.com/authorize";
  const RESPONSE_TYPE = "token";

  const logout = () => {
    setToken(null);
    window.localStorage.removeItem("token");
  };

  const handleLogin = () => {
    const params = new URLSearchParams();
    params.append("client_id", CLIENT_ID);
    params.append("response_type", RESPONSE_TYPE);
    params.append("redirect_uri", REDIRECT_URI);
    params.append(
      "scope",
      "user-read-private user-read-email user-library-read user-follow-read user-top-read user-modify-playback-state",
    );

    const url = `${AUTH_ENDPOINT}?${params.toString()}`;

    // Open Spotify login in a new window or redirect to the login URL
    window.open(url, "_blank");
  };

  const handleTokenFromCallback = () => {
    // Extract the token from the URL hash
    const urlParams = new URLSearchParams(window.location.hash.substr(1));
    const newToken = urlParams.get("access_token");

    if (newToken) {
      setToken(newToken);
      window.localStorage.setItem("token", newToken);
    }
  };

  function RenderUserProfile() {
    if (!userProfile) {
      return <p>No user profile available.</p>;
    }
    return (
      <div>
        <p>Display Name: {userProfile.display_name}</p>
        <p>User URL: {userProfile.external_urls.spotify}</p>
        <p>URI: {userProfile.uri}</p>
        <p>Total Followers: {userProfile.followers.total}</p>
      </div>
    );
  }

  const RenderTopArtists = () => {
    return topArtists.map((artist) => <div key={artist.id}>{artist.name}</div>);
  };

  // Check for token in the URL hash when component mounts
  React.useEffect(() => {
    handleTokenFromCallback();
  }, []);

  React.useEffect(() => {
    console.log("Token:", token);

    if (token) {
      fetch(`http://localhost:5000/getUserProfile?token=${token}`)
        .then((res) => res.json())
        .then((data) => setUserProfile(data.profile))
        .then(console.log("got user profile"));
    }
  }, [token]);

  React.useEffect(() => {
    console.log("Token:", token);
    if (token) {
      fetch(`http://localhost:5000/getTopItems?token=${token}`)
        .then((res) => res.json())
        .then((data) => {
          console.log("top artists: ", data);
          setTopArtists(data.topItems);
        })
        .catch((error) => console.error("Error fetching top items:", error));
    }
  }, [token]);

  const buttonStyle =
    "bg-white text-black py-2 px-4 rounded focus:outline-none focus:shadow-outline";

  return (
    <div className="flex flex-col items-center justify-center h-screen">
      {!token && (
        <button
          type="button"
          className={`mt-4 ${buttonStyle}`}
          onClick={handleLogin}
        >
          Login to Spotify
        </button>
      )}

      {token && (
        <div className="mt-4">
          <button type="button" className={`${buttonStyle}`} onClick={logout}>
            Logout
          </button>
        </div>
      )}

      {token && userProfile != null && (
        <div>User Profile:{RenderUserProfile()}</div>
      )}
      {token && <div>Your Top Artists:{RenderTopArtists()}</div>}
    </div>
  );
}

export default HomeContent;
