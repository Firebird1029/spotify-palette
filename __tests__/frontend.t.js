import React from "react";
import { render } from "@testing-library/react";

// Import the component to be tested
import HomePage from "../src/app/Index";

// https://github.com/vercel/next.js/discussions/58994
jest.mock("next/navigation", () => ({
  useRouter() {
    return {
      route: "/",
      pathname: "",
      query: "",
      asPath: "",
    };
  },
}));

// Jest snapshot test
describe("Index Component", () => {
  it("matches the snapshot", () => {
    const { asFragment } = render(<HomePage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
