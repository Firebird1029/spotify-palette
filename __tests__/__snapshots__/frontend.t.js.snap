// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index Component matches the snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="font-koulen"
    >
      Welcome to Uni.fy!
    </h1>
    <button
      class="border rounded-full px-6 py-2 text-lg font-koulen"
      type="button"
    >
      Log In
    </button>
    <button
      class="border rounded-full px-6 py-2 text-lg font-koulen"
      type="button"
    >
      Enter Code
    </button>
  </div>
</DocumentFragment>
`;
